slides/lectures/resources/bitwise/c.md

```diff
@@ -0,0 +1,119 @@
+# Bitwise Ops - in C
+
+## Set the `1` on position `bit` of `register` in C
+
+```c {all|2|3|4|1,5,6}
+unsigned int set_bit(unsigned int register_value, unsigned char bit) {
+    // assume register_value is 0b1000, bit is 2
+    //   1 << 2 is 0b0100
+    //   0b1000 | 0b0100 is 0b1100
+    return register_value | (1 << bit);
+}
+```
+
+##  Set multiple bits
+
+```c
+unsigned int set_bits(unsigned int register_value, unsigned int bits) {
+    // assume register_value is 0b1000, bits is 0b0111
+    //   0b1000 | 0b0111 is 0b1111
+    return register_value | bits;
+}
+```
+
+---
+
+# (test code)
+
+```c
+
+#include <stdio.h>
+
+void print_binary(unsigned int num) { 
+    for (int i = sizeof(num) * 8 - 1; i >= 0; i--) {
+        printf("%c", (num & (1 << i)) ? '1' : '0');
+    }
+    printf("\n"); //prints "num" number in binary format
+}
+
+unsigned int set_bits(unsigned int register_value, unsigned int bits) {
+    return register_value | bits;
+}
+
+int main() {
+    unsigned int reg = 0b1000;
+    unsigned int bits_to_set = 0b0011;
+
+    unsigned int result = set_bits(reg, bits_to_set);
+
+    printf("Register before: ");    print_binary(reg);
+    printf("Bits to set:     ");    print_binary(bits_to_set);
+    printf("Result after:    ");    print_binary(result);
+    return 0;
+}
+
+```
+
+
+---
+
+# Combined operation for value extraction in C
+
+```c
+#include <stdio.h>
+void print_binary(const char *label, unsigned int num) {
+    printf("%s: ", label);
+    for (int i = 31; i >= 0; i--) {
+        printf("%c", (num & (1 << i)) ? '1' : '0');
+    }
+    printf("\n");
+}
+
+int main() {
+    unsigned int large_id = 0b11001010111111000000111101101101;
+    unsigned int mask = 0b00000000000000000000111111111111;
+    unsigned int extracted_bits = (large_id >> 20) & mask;
```

Comment:
Code works. But why is the the large number shifted by 20 instead modifying/shifting the mask?

Reply:
I see no issue with the way the extracted bits are computed. But for me, it was more intuitive to see the outputted extracted bits as an overlap between large_id and mask.